working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+private name `_n`, and a constant variable named `_n` is added to the
+top-level of the current library. *(This means that we can move the
+parameter declarations including the default value without changing its
+meaning.)*
+
+Next, _k_ is a constant constructor iff the keyword `const` occurs just
+before the class name in the header of _D_, or _D_ is an `enum`
+declaration.
+
+If the class name `C` in _D_ is followed by `.id` where `id` is an
+identifier then _k_ has the name `C.id`. If it is followed by `.new` then
+_k_ has the name `C`. If the class name is not followed by `.` then _k_ has
+the name `C`.
+
+If it exists, _D2_ omits the part derived from `'.' <identifierOrNew>` that
+follows the class name in _D_.
+
+_D2_ omits the formal parameter list _L_ that follows the class name and
+possibly `.id` in _D_.
+
+The formal parameter list _L2_ of _k_ is identical to _L_, except that each
+formal parameter is processed as follows.
+
+In particular, the formal parameters in _L_ and _L2_ occur in the same
+order, and mandatory (respectively optional) positional parameters remain
+mandatory (respectively optional, with the same default value, if any), and
+named parameters preserve the name and the property of being `required`.
+
+- An initializing formal parameter *(e.g., `this.x`)* is copied from _L_ to
+  _L2_ without changes.
+- A super parameter is copied from _L_ to _L2_ without changes.
+- A formal parameter of the form `T p` or `final T p` where `T` is a type
+  and `p` is an identifier is replaced in _L2_ by `this.p`. Also, an
+  instance variable declaration of the form `T p;` or `final T p;` is added
+  to _D2_. The instance variable has the modifier `final` if the parameter
+  in _L_ is `final`, or _D_ has the modifier `inline` or is an `enum`
+  declaration, or the modifier `const` occurs just before the class name in
+  _D_.  In all cases, if `p` has the modifier `covariant` then this
+  modifier is removed from the parameter in _L2_, and it is added to the
+  instance variable declaration named `p`.
+
+Finally, _k_ is added to _D2_.
+
+### Discussion
+
+It could be argued that primary constructors should support arbitrary
+superinvocations using the specified superclass:
+
+```dart
+class B extends A { // OK.
+  B(int a): super(a);
+}
+
+class B(int a) extends A(a); // Could be supported, but isn't!
+```
+
+There are several reasons why this is not supported. First, primary constructors
+should be small and easy to read. Next, it is not obvious how the
+superconstructor arguments would fit into a mixin application (e.g., when the
+superclass is `A with M1, M2`), or how readable it would be if the
+superconstructor is named (`class B(int a) extends A.name(a);`). For instance,
+would it be obvious to all readers that the superclass is `A` and not `A.name`,
+and that all other constructors than the primary constructor will ignore the
+implied superinitialization `super.name(a)` and do their own thing (which might
+be implicit).
```

Comment:
We could also allow `extends A.name` with no arguments, just to choose where `super.x` parameters are forwarded.

Parsing-wise we then have to decide whether `id1.id2` is a prefix + class-name or a class-name+constructor-name. (And accept `id1.id2.id3` for the prefix-imported named constructor.)

Probably not an problem, it's only necessary to understand the identifiers at the point where we begin resolving identifiers, and there we can see what the identifiers point to.

Reply:
> We could also allow `extends A.name`

Yes, certainly. I expect that when this document is landed in 'working', discussions about lots of subtopics can start, as usual, and this would be one relevant area to discuss.